### ISSUE TITLE
`-fail-level` not respecting configured runner level (in e.g.: `reviewdog.yml`) (not sure if a bug or not)

### ISSUE THREAD

**Issue by Drowze** [reviewdog/reviewdog#1937]
Hello 👋 

Just ran on this today - it seems that the new `-fail-level` argument is not respecting the runner level that was configured in `.reviewdog.yml` config file. I'm not sure if this is expected or not tbh, but seems like a changed behaviour from the last version we were using (version v0.18.1 with `-reporter=github-pr-check -fail-on-error=true`)

Reproducing it:

<details>
<summary>Click to see all -fail-level=<level>+"runner level" combinations:</summary>

```yaml
# .reviewdog.yml
runner:
  i_runner:
    cmd: |
      ruby -e '$stdout.puts "foo.rb:42: example"'
    errorformat: [ '%f:%l: %m' ]
    level: info

  i_runner_i:
    cmd: |
      ruby -e '$stdout.puts "foo.rb:42:I: example"'
    errorformat: [ '%f:%l:%t: %m' ]
    level: info

  i_runner_w:
    cmd: |
      ruby -e '$stdout.puts "foo.rb:42:W: example"'
    errorformat: [ '%f:%l:%t: %m' ]
    level: info

  i_runner_e:
    cmd: |
      ruby -e '$stdout.puts "foo.rb:42:E: example"'
    errorformat: [ '%f:%l:%t: %m' ]
    level: info

  w_runner:
    cmd: |
      ruby -e '$stdout.puts "foo.rb:42: example"'
    errorformat: [ '%f:%l: %m' ]
    level: warning

  w_runner_i:
    cmd: |
      ruby -e '$stdout.puts "foo.rb:42:I: example"'
    errorformat: [ '%f:%l:%t: %m' ]
    level: warning

  w_runner_w:
    cmd: |
      ruby -e '$stdout.puts "foo.rb:42:W: example"'
    errorformat: [ '%f:%l:%t: %m' ]
    level: warning

  w_runner_e:
    cmd: |
      ruby -e '$stdout.puts "foo.rb:42:E: example"'
    errorformat: [ '%f:%l:%t: %m' ]
    level: warning

  e_runner:
    cmd: |
      ruby -e '$stdout.puts "foo.rb:42: example"'
    errorformat: [ '%f:%l: %m' ]
    level: error

  e_runner_i:
    cmd: |
      ruby -e '$stdout.puts "foo.rb:42:I: example"'
    errorformat: [ '%f:%l %t: %m' ]
    level: error

  e_runner_w:
    cmd: |
      ruby -e '$stdout.puts "foo.rb:42:W: example"'
    errorformat: [ '%f:%l %t: %m' ]
    level: error

  e_runner_e:
    cmd: |
      ruby -e '$stdout.puts "foo.rb:42:E: example"'
    errorformat: [ '%f:%l %t: %m' ]
    level: error
```

Testing all runner levels/fail-level combinations:
```bash
runners=(i_runner i_runner_i i_runner_w i_runner_e w_runner w_runner_i w_runner_w w_runner_e e_runner e_runner_i e_runner_w e_runner_e)

for runner in "${runners[@]}"; do
  runner_level="$(yq ".runner.$runner.level" .reviewdog.yml)"
  reviewdog -runners="$runner" -fail-level=info &>/dev/null
  exit_code=$?
  printf "%18s:%20s%25s%15s\n" "fail-level=info" "runner=$runner" "runner_level=$runner_level" "exit_code=$exit_code"
done

for runner in "${runners[@]}"; do
  runner_level="$(yq ".runner.$runner.level" .reviewdog.yml)"
  reviewdog -runners="$runner" -fail-level=warning &>/dev/null
  exit_code=$?
  printf "%18s:%20s%25s%15s\n" "fail-level=warning" "runner=$runner" "runner_level=$runner_level" "exit_code=$exit_code"
done

for runner in "${runners[@]}"; do
  runner_level="$(yq ".runner.$runner.level" .reviewdog.yml)"
  reviewdog -runners="$runner" -fail-level=error &>/dev/null
  exit_code=$?
  printf "%18s:%20s%25s%15s\n" "fail-level=error" "runner=$runner" "runner_level=$runner_level" "exit_code=$exit_code"
done
```

Output:
```
   fail-level=info:     runner=i_runner        runner_level=info    exit_code=1
   fail-level=info:   runner=i_runner_i        runner_level=info    exit_code=1
   fail-level=info:   runner=i_runner_w        runner_level=info    exit_code=1
   fail-level=info:   runner=i_runner_e        runner_level=info    exit_code=1
   fail-level=info:     runner=w_runner     runner_level=warning    exit_code=1
   fail-level=info:   runner=w_runner_i     runner_level=warning    exit_code=1
   fail-level=info:   runner=w_runner_w     runner_level=warning    exit_code=1
   fail-level=info:   runner=w_runner_e     runner_level=warning    exit_code=1
   fail-level=info:     runner=e_runner       runner_level=error    exit_code=1
   fail-level=info:   runner=e_runner_i       runner_level=error    exit_code=0
   fail-level=info:   runner=e_runner_w       runner_level=error    exit_code=0
   fail-level=info:   runner=e_runner_e       runner_level=error    exit_code=0
fail-level=warning:     runner=i_runner        runner_level=info    exit_code=1
fail-level=warning:   runner=i_runner_i        runner_level=info    exit_code=0
fail-level=warning:   runner=i_runner_w        runner_level=info    exit_code=1
fail-level=warning:   runner=i_runner_e        runner_level=info    exit_code=1
fail-level=warning:     runner=w_runner     runner_level=warning    exit_code=1
fail-level=warning:   runner=w_runner_i     runner_level=warning    exit_code=0
fail-level=warning:   runner=w_runner_w     runner_level=warning    exit_code=1
fail-level=warning:   runner=w_runner_e     runner_level=warning    exit_code=1
fail-level=warning:     runner=e_runner       runner_level=error    exit_code=1
fail-level=warning:   runner=e_runner_i       runner_level=error    exit_code=0
fail-level=warning:   runner=e_runner_w       runner_level=error    exit_code=0
fail-level=warning:   runner=e_runner_e       runner_level=error    exit_code=0
  fail-level=error:     runner=i_runner        runner_level=info    exit_code=1
  fail-level=error:   runner=i_runner_i        runner_level=info    exit_code=0
  fail-level=error:   runner=i_runner_w        runner_level=info    exit_code=0
  fail-level=error:   runner=i_runner_e        runner_level=info    exit_code=1
  fail-level=error:     runner=w_runner     runner_level=warning    exit_code=1
  fail-level=error:   runner=w_runner_i     runner_level=warning    exit_code=0
  fail-level=error:   runner=w_runner_w     runner_level=warning    exit_code=0
  fail-level=error:   runner=w_runner_e     runner_level=warning    exit_code=1
  fail-level=error:     runner=e_runner       runner_level=error    exit_code=1
  fail-level=error:   runner=e_runner_i       runner_level=error    exit_code=0
  fail-level=error:   runner=e_runner_w       runner_level=error    exit_code=0
  fail-level=error:   runner=e_runner_e       runner_level=error    exit_code=0
```

I believe reviewdog is ignoring the _runner level_ for calculating the exit code, and only taking into consideration the level parsed out of the runners' messages.

</details>

Minimal repro:
```yaml
#.reviewdog.yml
runner:
  error_level_runner:
    cmd: 'true'
    errorformat:
      - '%f:%l: %m'
    level: error

  info_level_runner:
    cmd: |
      ruby -e '$stdout.puts "foo.rb:42: example info"'
    errorformat:
      - '%f:%l: %m'
    level: info
```

```
$ reviewdog -fail-level=error
time=2024-09-20T14:16:29.796-03:00 level=INFO msg="reviewdog: [start] runner=error_level_runner"
time=2024-09-20T14:16:29.797-03:00 level=INFO msg="reviewdog: [start] runner=info_level_runner"
time=2024-09-20T14:16:29.837-03:00 level=INFO msg="reviewdog: [finish] runner=error_level_runner"
time=2024-09-20T14:16:29.960-03:00 level=INFO msg="reviewdog: [finish] runner=info_level_runner"
foo.rb:42: [info_level_runner] example info
reviewdog: found at least one issue with severity greater than or equal to the given level: error
```

**Comment by Drowze** [comment]
🎉 workaround found!

Finally came back to this a couple years later and found a workaround for our usecase, where we run multiple runners and don't want our CI to due to `info`/`warn` runners.

_...scroll down for a tl;dr..._

### For context, investigation and random thoughts....

As I was working on refactoring our CI and reviewdog usage, I stumbled again on this issue. Our usecase is simple: run multiple runners but do NOT exit with a non-zero status code if info/warning runners fail.

Here's a simplified version of our `.reviewdog.yml`:
```yaml
# NOTE: `rubocop` & `brakeman` are ruby linters with built-in support in reviewdog
runner:
  rubocop:
    cmd: rubocop
    format: rubocop
    level: error
  brakeman:
    cmd: brakeman -q -f tabs
    format: brakeman
    level: warning
```

Local usage:
```bash
$ reviewdog -runners=rubocop,brakeman -fail-level=error
```

**Expected behavior:** do not fail when only brakeman messages are reported. Always fail when any rubocop message is reported.
**Actual behavior:** always failing when any brakeman/rubocop message is reported.

To investigate it, let's have a look at a sample output for `brakeman -q -f tabs`:
```
/home/Users/Drowze/app/models/my_file.rb	35	Remote Code Execution	General	`Oj.load` called with parameter value near line 35: Oj.load(params[:filters], :symbol_keys => true)	High
/home/Users/Drowze/config/setup/my_other_file_.rb	8	Redirect	General	Possible unprotected redirect near line 8: redirect_to(Links.find_by(:token => params[:token]).target_url, :allow_other_host => true)	Weak
/home/Users/Drowze/app/controllers/some_other_file.rb	128	SQL Injection	General	Possible SQL injection near line 128: user.joins("JOIN users ON subscriptions.subject_id = users.id AND usesrs.type = '#{type}'")	Medium
/home/Users/Drowze/lib/last_one_i_promise.rb	440	Remote Code Execution	General	Unsafe reflection method `constantize` called on model attribute near line 440: UserAlias.name.constantize	Medium
```

Taking a good look at the output above, you'll notice it does not include the messages' severity level (note: `High`/`Medium`/`Weak` there concerns to the message _confidence level_ - a different brakeman concept).

With that in mind I've investigated and played a bit with `reviewdog`'s errorformats/runner levels/fail-level and I am now confident that:
* The `level` setting in the `.reviewdog.yml` config has nothing to do with its messages severity level: <ins>**if the message has no severity, it is ALWAYS considered of `error` severity**</ins>.
* As a consequence, as of today <ins>**it is not possible to use the builtin formats (like `-f=brakeman`) for messages that lack a severity level and consider those messages as info/warning levels**</ins>; i.e.: when a message does not include its level, it will always be considered as error level and always throw an error when `-fail-level={info,warning,error}` is specified.

Here's a demonstration for 
```yaml
# .reviewdog.yml
runner:
  brakeman:
    cmd: brakman -q -f tabs
    format: brakeman
    # Clarifying as this also confuses me:
    # * reviewdog understands this as "only report messages with severity of warning or higher"
    # * reviewdog does NOT understand it as "report all messages as warnings"
    # * reviewdog does NOT understand it as "report all messages without a severity as warnings"
    #
    # I.e.: regardless of this setting, all messages without a severity will ALWAYS
    #   be reported as an error; so reviewdog will ALWAYS exit with a non-zero exit code
    #   when using `-fail-level={error,warning,info}` when a message without a severity is reported.
    level: warning
```

```bash
# suppressing non-relevant output for brevity..

$ reviewdog -runners="brakeman"
time=2026-08-04T10:47:10.880-03:00 level=INFO msg="reviewdog: [start] runner=brakeman"
time=2026-08-04T10:47:10.922-03:00 level=INFO msg="reviewdog: [finish] runner=brakeman\terror=exit status 3"
...

$ echo "$?"
0

# note the exit code is now non-zero, as the brakeman messages were considered errors
$ reviewdog -runners="brakeman" -fail-level=error
time=2026-08-04T10:49:16.104-03:00 level=INFO msg="reviewdog: [start] runner=brakeman"
time=2026-08-04T10:49:16.132-03:00 level=INFO msg="reviewdog: [finish] runner=brakeman\terror=exit status 3"
...
reviewdog: found at least one issue with severity greater than or equal to the given level: error

$ echo "$?"
1
```

### Workaround

Looking at the [errorformat vim documentation](https://vimdoc.sourceforge.net/htmldoc/quickfix.html#errorformat) I found these:
```
	%E		start of a multi-line error message
	%W		start of a multi-line warning message
	%I		start of a multi-line informational message
```

After testing with it a bit it's perfectly possible to use those to force a severity level to messages, effectively fixing our problem. So, in the case of brakeman, [according to here](https://github.com/reviewdog/errorformat/blob/13bff69235f30fd8b35c60151582584a7d81a50a/fmts/ruby.go#L9) its errorformat is `%f%*\s%l%*\s%m`. Updating our demo configuration with what we've learned:
```yaml
runner:
  brakeman:
    cmd: brakeman -q -f tabs
    errorformat:
      - '%I%f%*\s%l%*\s%m' # prefix with `%I` so all error messages are forced to `info` severity level
    level: warning
```

Testing with our new config:
```bash
# suppressing non-relevant output for brevity..
$ reviewdog -runners=brakeman -fail-level=info
time=2026-08-04T12:00:12.328-03:00 level=INFO msg="reviewdog: [start] runner=brakeman"
time=2026-08-04T12:00:12.357-03:00 level=INFO msg="reviewdog: [finish] runner=brakeman\terror=exit status 3"
...
reviewdog: found at least one issue with severity greater than or equal to the given level: info

$ echo "$?"
1

$ reviewdog -runners=brakeman -fail-level=warning
time=2026-08-04T12:01:47.464-03:00 level=INFO msg="reviewdog: [start] runner=brakeman"
time=2026-08-04T12:01:47.494-03:00 level=INFO msg="reviewdog: [finish] runner=brakeman\terror=exit status 3"
...

$ echo "$?"
0

$ reviewdog -runners=brakeman -fail-level=error
time=2026-08-04T12:01:49.464-03:00 level=INFO msg="reviewdog: [start] runner=brakeman"
time=2026-08-04T12:01:49.494-03:00 level=INFO msg="reviewdog: [finish] runner=brakeman\terror=exit status 3"
...

$ echo "$?"
0
```

### tl;dr

To force info/warning severity level to a message parsed by Reviewdog, simply prefix `%I` / `%W` to its errorformat. It's not possible to force a severity level on messages parsed using built-in formats though (afaik).
Leveraging that we should be able to use `-fail-level=...` with mixed severity runners.

**Comment by Drowze** [comment]
I've asked Copilot to research and help writing a detailed explanation of the `-fail-level` vs per-runner `level:` behavior (under v0.21.0). You can find it below - hope it can be useful!

---

**TL;DR: `-fail-level` and the per-runner `level:` config are two completely disconnected mechanisms. `-fail-level` only ever looks at each `rdf.Diagnostic`'s own parsed `Severity` field — it never reads `runner.Level`. `level:` is only used for GitHub annotation fallback/conclusion, never for the process exit code (except via a second, separate code path used by GitHub Apps auth — see below).**

### 1. `-fail-level` compares only `rdf.Diagnostic.Severity`, never `runner.Level`

- [`reviewdog.go:119`](https://github.com/reviewdog/reviewdog/blob/v0.21.0/reviewdog.go#L119): `shouldFail = shouldFail || w.failLevel.ShouldFail(check.Diagnostic.GetSeverity())` — the only place exit-worthiness is computed. Note it's driven entirely by `Diagnostic.GetSeverity()`, with no reference to any runner/tool-level config.
- [`fail_level.go:67-73`](https://github.com/reviewdog/reviewdog/blob/v0.21.0/fail_level.go#L67-L73):
  ```go
  func (failLevel FailLevel) ShouldFail(severity rdf.Severity) bool {
      if failLevel == FailLevelDefault || failLevel == FailLevelNone {
          return false
      }
      minSeverity := failLevel.minSeverity()
      return minSeverity == rdf.Severity_UNKNOWN_SEVERITY || severity <= minSeverity
  }
  ```
  **Bug within a bug**: `Severity_UNKNOWN_SEVERITY == 0` and `Severity_ERROR/WARNING/INFO == 1/2/3` (proto/rdf/reviewdog.pb.go:41-44), so `severity <= minSeverity` is trivially true whenever a diagnostic's severity is `UNKNOWN` (0), for *any* `-fail-level` other than `none`. **Any diagnostic reviewdog can't map to a known severity always fails the run**, regardless of `-fail-level`.
- [`project/run.go:97`](https://github.com/reviewdog/reviewdog/blob/v0.21.0/project/run.go#L97) even says so explicitly in a comment: `RunAndParse(ctx, conf, runners, "", teeMode) // Level is not used.`
- [`project/run.go:119`](https://github.com/reviewdog/reviewdog/blob/v0.21.0/project/run.go#L119) and [`:122`](https://github.com/reviewdog/reviewdog/blob/v0.21.0/project/run.go#L122): `result.Level` (i.e. your runner's `level:` value) is passed only to `ncs.SetTool(toolname, result.Level)` (a `CommentService` method, used purely for annotation rendering) — the *same* global `failLevel` is passed to `RunFromResult` for **every** runner regardless of its individual `level:`.

### 2. `level:`/`-level` only affects GitHub annotation fallback + Check "conclusion", never the exit code

- [`service/github/check.go:312`](https://github.com/reviewdog/reviewdog/blob/v0.21.0/service/github/check.go#L312) `annotationLevel()` uses `c.Diagnostic.Severity` per-annotation, falling back to `ch.Level` ([`:325`](https://github.com/reviewdog/reviewdog/blob/v0.21.0/service/github/check.go#L325) `reqAnnotationLevel()`) only when severity is unknown.
- [`service/github/check.go:257-269`](https://github.com/reviewdog/reviewdog/blob/v0.21.0/service/github/check.go#L257-L269):
  ```go
  func (ch *Check) conclusion(annotations []*github.CheckRunAnnotation) string {
      checkResult := "success"
      if ch.Level != "" {
          // Level takes precedence when configured (for backwards compatibility)
          if len(annotations) == 0 { return checkResult }
          switch strings.ToLower(ch.Level) {
          case "info", "warning":
              return "neutral"
          }
          return "failure"
      }
      ...
  }
  ```
  This is the GitHub Check Run's own `conclusion` field (what shows pass/fail in the GitHub UI for that check) — **entirely separate from reviewdog's process exit code**. When `level:` is set, it overrides per-diagnostic severity for this field only.

### 3. A *second*, separate bug: `-fail-level` is ignored completely when using GitHub Apps auth (`REVIEWDOG_TOKEN`)

- [`cmd/reviewdog/main.go:320-322`](https://github.com/reviewdog/reviewdog/blob/v0.21.0/cmd/reviewdog/main.go#L320-L322): when reporter is `github-check`/`github-pr-check` and `!skipDoghouseServer()` (i.e., using `REVIEWDOG_TOKEN` GitHub App auth instead of `REVIEWDOG_GITHUB_API_TOKEN`), reviewdog takes a **completely different path**: `runDoghouse()` → `postResultSet()`, which never calls `RunFromResult`/`fail_level.go` at all.
- [`cmd/reviewdog/doghouse.go:115`](https://github.com/reviewdog/reviewdog/blob/v0.21.0/cmd/reviewdog/doghouse.go#L115) sends `result.Level` (your runner's config) to the doghouse server, which computes `res.Conclusion` server-side (mirroring the same `level`-overrides-severity logic as `Check.conclusion()` above).
- [`cmd/reviewdog/doghouse.go:145-146`](https://github.com/reviewdog/reviewdog/blob/v0.21.0/cmd/reviewdog/doghouse.go#L145-L146): `if opt.failOnError && (res.Conclusion == "failure") { return error }` — this only checks the **deprecated** `-fail-on-error` boolean, not `-fail-level` at all.
- [`cmd/reviewdog/main.go:1048-1061`](https://github.com/reviewdog/reviewdog/blob/v0.21.0/cmd/reviewdog/main.go#L1048-L1061) `failLevel(opt)` shows `-fail-on-error=true` maps to `FailLevelAny` for most reporters but `FailLevelError` specifically for `github-check`/`github-pr-check` — but this mapped value is **never actually consulted** by the doghouse path in point above; only the raw boolean is.
- This is what issue **#1407** independently discovered and traced to the same lines.

### Practical workaround (verified working)

Force severity at the errorformat level instead of relying on `level:`, using `%E`/`%W`/`%I` prefixes (not `%A` + `%t`, since `%t` overrides the prefix-implied type — see `errorformat` lib's `errorformat.go` around the `EWI` prefix handling):
```yaml
runner:
  rubocop:
    cmd: bundle exec rubocop
    errorformat:
      - '%E%f:%l:%c: %*[RCWEF]: %m'   # force ALL rubocop offenses to ERROR severity
      - '%Z%p^%#'
      - '%C%.%#'
      - '%-G%.%#'
  brakeman:
    cmd: bundle exec brakeman -q -f tabs
    errorformat:
      - '%W%f%*\s%l%*\s%m'            # force ALL brakeman findings to WARNING severity
      - '%Z'
```
With this, a single `-fail-level=error` correctly fails only on rubocop findings and never on brakeman, because both runners' diagnostics now carry real, forced severities instead of relying on the ignored `level:` field.

### Suggested fixes for maintainers
1. Either make `-fail-level` respect the per-runner `level:` as a floor/override (so `level: warning` truly means "these findings can never fail the build past `warning`"), or remove `level:`'s implied connection to failure semantics from the docs entirely and document it as **annotation-only**.
2. Wire `-fail-level` into the doghouse (`REVIEWDOG_TOKEN`) code path (`postResultSet`) instead of only the deprecated `-fail-on-error` boolean, so behavior is consistent regardless of auth method.
